pkg/relay/ethereum/signer.go

```diff
@@ -1,20 +1,69 @@
 package ethereum
 
 import (
+	"fmt"
 	"math/big"
 
-	"github.com/cosmos/gogoproto/proto"
+	"github.com/hyperledger-labs/yui-relayer/log"
+	"github.com/hyperledger-labs/yui-relayer/signer"
 	"github.com/ethereum/go-ethereum/common"
+	gethcrypto "github.com/ethereum/go-ethereum/crypto"
 	gethtypes "github.com/ethereum/go-ethereum/core/types"
 )
 
-type SignerConfig interface {
-	proto.Message
-	Build(chainID *big.Int) (Signer, error)
-	Validate() error
+type EthereumSigner struct {
+	bytesSigner signer.Signer
+	gethSigner gethtypes.Signer
+	addressCache common.Address
+	logger *log.RelayLogger
 }
 
-type Signer interface {
-	Sign(common.Address, *gethtypes.Transaction) (*gethtypes.Transaction, error)
-	Address() common.Address
+func NewEthereumSigner(bytesSigner signer.Signer, chainID *big.Int) (*EthereumSigner, error) {
+	pkbytes, err := bytesSigner.GetPublicKey()
+	if err != nil {
+		return nil, fmt.Errorf("fail to get public key")
+	}
+
+	pk, err := gethcrypto.DecompressPubkey(pkbytes)
+	if err != nil {
+		return nil, fmt.Errorf("fail to decompress public key")
+	}
+
+	addr := gethcrypto.PubkeyToAddress(*pk)
+
+	gethSigner := gethtypes.LatestSignerForChainID(chainID)
+
+	return &EthereumSigner {
+		bytesSigner: bytesSigner,
+		gethSigner: gethSigner,
+		addressCache: addr,
+		logger: nil,
+	}, nil
+}
+
+func (s *EthereumSigner) SetLogger(logger *log.RelayLogger) {
+	s.logger = logger
+}
```

Comment:
After calling `SetLogger` and finishing the processing, we need to reset the logger, right?
```
origLogger := c.ethereumSigner.GetLogger()
c.ethereumSigner.SetLogger(newLogger)

<... some processing ...>

c.ethereumSigner.SetLogger(origLogger)
```
or
```
c.ethereumSigner.PushLogger(newLogger)

<... some processing ...>

defer c.ethereumSigner.PopLogger()
```

In any case, we should define an interface that clearly indicates the section where the configured logger is used.

Reply:
This EthereumSigner use logger only in Sign function and no need to have dedicate logger instance(such like use different output log file).
How abount Sign function get logger or ctx which has logger as an arument?

```
func (s *EthereumSigner) Sign(address common.Address, tx *gethtypes.Transaction, logger *log.RelayLogger) (*gethtypes.Transaction, error) {
```

Reply:
@dai1975 Thank you. If the function signature of `EthereumSigner::Sign` could be changed, I would agree with the idea, but it can't.
The function signature must match go-ethereum's [SignerFn](https://github.com/ethereum/go-ethereum/blob/v1.14.6/accounts/abi/bind/base.go#L60), right?

Reply:
I see.
fixed by using `defer SetLogger(oldlogger)` pattern in tx.go